discovery-provider/src/tasks/index_plays.py

```diff
@@ -0,0 +1,131 @@
+import logging
+from urllib.parse import urljoin
+import requests
+import dateutil.parser
+import datetime
+from sqlalchemy import func, desc
+from src.models import Play
+from src.tasks.celery_app import celery
+
+logger = logging.getLogger(__name__)
+
+# The number of listens to get per request to identity
+REQUEST_LISTENS_LIMIT = 5000
+
+# Retrieve the play counts from the identity service
+# NOTE: indexing the plays will eventually be a part of `index_blocks`
+
+
+def get_track_plays(self, db):
+    with db.scoped_session() as session:
+        # Get the most retrieved play date in the db to use as an offet for fetching
+        # more play counts from identity
+        most_recent_play_date = session.query(
+            Play.updated_at
+        ).order_by(
+            desc(Play.updated_at),
+            desc(Play.id)
+        ).first()
+        if most_recent_play_date == None:
+            # Make the date way back in the past to get the first play count onwards
+            most_recent_play_date = datetime.datetime(
+                2000, 1, 1, 0, 0).timestamp()
+        else:
+            most_recent_play_date = most_recent_play_date[0].timestamp()
+
+        # Create and query identity service endpoint for track play counts
+        identity_url = update_play_count.shared_config['discprov']['identity_service_url']
+        params = {'startTime': most_recent_play_date,
+                  'limit': REQUEST_LISTENS_LIMIT}
+        identity_tracks_endpoint = urljoin(identity_url, 'listens/bulk')
+
+        track_listens = {}
+        try:
+            resp = requests.get(identity_tracks_endpoint, params=params)
+            track_listens = resp.json()
+        except Exception as e:
+            logger.error(
+                f'Error retrieving track play counts - {identity_tracks_endpoint}, {e}'
+            )
+
+        # Insert a new row for each count instance in the plays table
+        plays = []
+        if 'listens' in track_listens:
+            for listen in track_listens['listens']:
+                if 'userId' in listen and listen['userId'] != None:
+                    # If the userId is present, query for exist plays and only
+                    # insert new plays for the difference
+                    user_track_play_count = session.query(
```

Comment:
only concern i have here is the volume of db calls we'd be making for this. our connectin pool size is 10 and i want to make sure we don't saturate all the connections doing this @jowlee @raymondjacobson

Reply:
Maybe we should loop over track listens first and do a single bulk query, although this one seems real weird cause it's requires userId/trackId combo per query. thoughts on how to reduce/ what we should do @raymondjacobson ?

Reply:
let's merge this into feature_api and open a second ticket to explore this

Reply:
I think dheeraj is right though, we can do a bulk query for everything we would need and do the work in memory in python and then do a bulk write back.